### PR TITLE
bug/Timers being lost on shutdown due to a race condition in nats

### PR DESCRIPTION
**Pull request by Iaskebba**

### CODE REVIEW

File: lib/timer.js [hunk 1]
```diff
@@ -201,26 +201,33 @@ timers.create(id, options, (err) => {
...
+          , payload.timeout - elapsed
+          , payload.callback.method
+          , payload.callback.uri
+          , payload.data
+          , id
+          , this
+        ).unref();
+        this.set(id, data);
+        cb(null, id);
+      });
```
Comment (Iaskebba) [inline lib/timer.js]: Sometimes 201s were being returned before the timer was actually put into persistent storage due to the noop

File: lib/timer.js [hunk 2]
```diff
@@ -418,13 +425,17 @@ timers.success('2e2f6dad-9678-4caf-bc41-8e62ca07d551', error)
...
       this.nats.publish('skyring:node', JSON.stringify({
         node: this[kNode]
       , type: EVENT_STATUS.SHUTDOWN
-      }), noop)
-      setImmediate(this.nats.quit, cb)
+      }), function() {
+        client.flush(function() {
+          setImmediate(client.quit, cb);
+        });
+      });
```
Comment (Iaskebba) [inline lib/timer.js]: Flushing out the nats before closing fixes messages from being lost.

File: lib/timer.js [hunk 3]
```diff
@@ -201,26 +201,33 @@ timers.create(id, options, (err) => {
       debug('setting timer', id);
       //TODO(esatterwhite):
       // what should happen if leveldb fails.
-      if (err) console.error(err);
+      if (err) {
+        console.error(err);
+        cb(err, null);
```
Comment (esatterwhite) [inline lib/timer.js]: This will have to `return cb(err, null)` or this will allow a double callback

File: lib/timer.js [hunk 4]
```diff
@@ -201,26 +201,33 @@ timers.create(id, options, (err) => {
...
-      , payload.data
-      , id
-      , this
-      ).unref();
-      this.set( id, data );
-      cb(null, id);
+      }), (err) => {
+        if (err) {
+          console.error(err);
+          cb(err, null);
```
Comment (esatterwhite) [inline lib/timer.js]: There isn't really a need to capture an error on this. it is for monitoring. 
this specific call failing shouldn't effect anything else.

lines 208-214 could  be removed and this should behave the same.

File: lib/timer.js (hunk 1 above)
Comment (esatterwhite) [reply lib/timer.js]: I'm not sure the two things were directly related. The the nats.publish is just for external monitoring. It doesn't have any bearing on the  actual behavior of the server.

File: lib/timer.js (hunk 2 above)
Comment (esatterwhite) [reply lib/timer.js]: Yep, this makes sense 👍

File: lib/timer.js (hunk 1 above)
Comment (esatterwhite) [reply lib/timer.js]: This still seems like a race between the callback being called and the write ack between mongo.
The timer was already being set after the leveldb write happened. All you've done here is push that down to after a nats call which is just for monitoring.

I wonder if it something with the default write concern on mongo or if there is something that can be passed to the driver...

**Pull request merged**